core/src/Cuda/Kokkos_Cuda_Instance.cpp

```diff
@@ -157,16 +165,20 @@ void cuda_stream_synchronize(
     Kokkos::Tools::Experimental::SpecialSynchronizationCases reason,
     const std::string &name) {
   Kokkos::Tools::Experimental::Impl::profile_fence_event<Kokkos::Cuda>(
-      name, reason, [&]() {  // TODO: correct device ID
-        KOKKOS_IMPL_CUDA_SAFE_CALL(cudaStreamSynchronize(stream));
+      name, reason, [&]() {
+        KOKKOS_IMPL_CUDA_SAFE_CALL(
+            (Impl::CudaInternal::singleton().cuda_stream_synchronize_wrapper(
+                stream)));
       });
 }
 
 void cuda_internal_error_throw(cudaError e, const char *name, const char *file,
                                const int line) {
   std::ostringstream out;
-  out << name << " error( " << cudaGetErrorName(e)
-      << "): " << cudaGetErrorString(e);
+  out << name << " error( "
+      << CudaInternal::singleton().cuda_get_error_name_wrapper<false>(e)
+      << "): "
+      << CudaInternal::singleton().cuda_get_error_string_wrapper<false>(e);
```

Comment:
Is there a particular reason not to set the device here?

Reply:
I did this since all the cudaAPI calls wrapped in `KOKKOS_IMPL_CUDA_SAFE_CALL()` will hit this (or below) if there is a runtime error, and we can't be guaranteed that the call that is erroring out could set device or not. For example if `cuda_get_device_count_wrapper<false>(...)` in line 271 of this file had an error, we would not get the message because before error was thrown, these functions would complain that the device was not initialized.

Reply:
Sounds reasonable.